### ISSUE TITLE
torch.Tensor.random_ is divergent from NumPy's np.random.random

### ISSUE THREAD

**Issue by mruberry** [pytorch/pytorch#50332]
[torch.Tensor.random_](https://pytorch.org/docs/master/tensors.html?highlight=random_#torch.Tensor.random_) is divergent from NumPy's [np.random.random](https://numpy.org/doc/stable/reference/random/generated/numpy.random.random.html?highlight=random#numpy.random.random), which is an alias for [np.random.random_sample](https://numpy.org/doc/stable/reference/random/generated/numpy.random.random_sample.html#numpy.random.random_sample). 

In short, PyTorch's torch.Tensor.random_ returns values of different dtypes from a dtype-dependent range, while NumPy's random.random_sample returns elements in the [0, 1) interval. 

This issue was first reported in https://github.com/pytorch/pytorch/issues/1875.

**Comment by mruberry** [comment]
I propose we don't fix this since NumPy does not have the inplace variant `random_` and `np.random.random` is an alias to the newer API.

Separately we should consider implementing a random module in PyTorch.

@rgommers, what do you think about this issue and a potential random module? How are the random number generator discussions going for the array API?

**Comment by rgommers** [comment]
NumPy has two random number generation APIs in the same `numpy.random` module. Currently >95% of code uses the legacy API. The problem is that it's not very good; the new API is both conceptually much better (improved seeding, no global state, good parallel generation facilities) and has multiple different generators with better performance (see https://numpy.org/devdocs/reference/random/performance.html).

This is probably also the reason that JAX does not have `numpy.random`, it's the only missing numpy-compat module. Their `jax.random` does look more like the new NumPy API than the legacy one. The key feature being no global state, and having to pass a PRNG instance to all functions. For notes on why, see https://github.com/google/jax/blob/master/design_notes/prng.md

[cupy.random](https://docs.cupy.dev/en/stable/reference/random.html) only implements the legacy NumPy API.

Then the PyTorch facilities: they're currently similar to the legacy NumPy API, but with only seeding and fork functionality in `torch.random`, and the rest of the functionality in the main namespace. Plus then an odd split between some functions, and a few more methods on `Tensor` (see note under https://pytorch.org/docs/master/torch.html#creation-ops). Far fewer than NumPy though. The generator PyTorch uses is `PCG64`, which is a good choice (it's also NumPy's new default), and certainly better than `MT19937`. Finally there's quasi-random sampling through `torch.quasirandom.SobolEngine`, which is a subset of what the new [scipy.stats.qmc](http://scipy.github.io/devdocs/reference/stats.qmc.html) module provides.

The above explains why we haven't thought about standardization much: most libraries diverge quite a bit, and the most common API is inferior so we don't want to enshrine it. 

The other thing there is that it's really hard to teach people to use one generator instance and thread that through all of their code. Having the `np.random.somefunc` utility functions is just easier and _seems_ to work (until it silently doesn't, see recent noise about `DataLoader` + `numpy.random`).

I think there would be a lot of value for PyTorch to adopt the new NumPy API; either that or do nothing, I would not recommend expanding `torch.random` to match the legacy NumPy API.

**Comment by mruberry** [comment]
From offline review: another possible solution would be to just remove random_
